crates/rooch-key/src/key_derive.rs

```diff
@@ -31,49 +41,296 @@ pub const DERVIATION_PATH_PURPOSE_SCHNORR: u32 = 44;
-        Ok((address, kp.into())) // Cast to KeyPair
+        Ok((address, kp.into()))
+    }
+
+    fn retrieve_key_pair(
+        &self,
+        encryption: &EncryptionData,
+        password: Option<String>,
+    ) -> Result<RoochKeyPair, RoochError> {
+        let is_verified = verify_password(password.clone(), encryption.hashed_password)
+            .expect("Verification failed for password");
+        if is_verified {
+            let private_key = decrypt_private_key(
+                &encryption.nonce,
+                &encryption.ciphertext,
+                &encryption.tag,
+                password,
+            )
+            .expect("Decryption failed for private key");
+            let kp: Ed25519KeyPair = Ed25519KeyPair::from(
+                Ed25519PrivateKey::from_bytes(&private_key)
+                    .map_err(|e| RoochError::SignatureKeyGenError(e.to_string()))?,
+            );
+            Ok(kp.into())
+        } else {
+            Err(RoochError::InvalidPasswordError(
+                "Password is invalid".to_owned(),
+            ))
+        }
+    }
+
+    // Encrypt the private key using ChaCha20Poly1305 and Argon2
+    fn encrypt_private_key(
+        &self,
+        private_key: Ed25519PrivateKey,
+        password: Option<String>,
+    ) -> Result<(Vec<u8>, Vec<u8>, Vec<u8>), RoochError> {
+        // 96-bits; unique per message
+        let nonce = ChaCha20Poly1305::generate_nonce(&mut OsRng);
+
+        // Derive the key material using nonce and password
+        let mut output_key_material = [0u8; 32]; // Can be any desired size
+        Argon2::default()
+            .hash_password_into(
+                password.unwrap_or("".to_owned()).as_bytes(),
+                &nonce,
+                &mut output_key_material,
+            )
+            .map_err(|e| RoochError::KeyConversionError(e.to_string()))?;
+
+        // Create a ChaCha20Poly1305 cipher with the key material from password
+        let cipher = ChaCha20Poly1305::new_from_slice(&output_key_material)
+            .map_err(|e| RoochError::KeyConversionError(e.to_string()))?;
+
+        // Encrypt the private key data to a ciphertext with a tag
+        let ciphertext_with_tag = match cipher.encrypt(&nonce, private_key.as_bytes()) {
+            Ok(ciphertext) => ciphertext,
+            Err(_) => {
+                return Err(RoochError::KeyConversionError(
+                    "Encryption failed".to_owned(),
+                ))
+            }
+        };
+
+        // Extract the ciphertext without the tag
+        let ciphertext = ciphertext_with_tag[..ciphertext_with_tag.len() - 16].to_vec();
+
+        // Extract the tag (last 16 bytes)
+        let tag = {
+            let start = ciphertext_with_tag.len() - 16;
+            let end = ciphertext_with_tag.len();
+            let mut tag = Vec::with_capacity(16);
+            tag.extend_from_slice(&ciphertext_with_tag[start..end]);
+            tag
+        };
```

Comment:
Why split the tag? or include the tag in the ciphertext? I can not find the use case of the tag.

Reply:
> Why split the tag? or include the tag in the ciphertext? I can not find the use case of the tag.

The tag is useless here, and why did I split it is cause the private key should be 32 bytes. The ciphertext is composed of 32 bytes encrypted text and 16 bytes tag thus 48 bytes. In order to achieve key derivations, i.e. parsing the encrypted cipertext to private keys and vice versa, it needs to be split.

Reply:
> [2.5](https://www.rfc-editor.org/rfc/rfc7539#section-2.5).  The Poly1305 Algorithm
> 
>    Poly1305 is a one-time authenticator designed by D. J. Bernstein.
>    Poly1305 takes a 32-byte one-time key and a message and produces a
>    16-byte tag.  This tag is used to authenticate the message.

https://www.rfc-editor.org/rfc/rfc7539

Reply:
Add a comment to the tag.